service/worker/sysworkflow/system_workflow.go

```diff
@@ -22,101 +22,337 @@ package sysworkflow
+	"github.com/uber/cadence/common/blobstore"
+	"github.com/uber/cadence/common/blobstore/blob"
 	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
 	"go.uber.org/cadence"
+	"go.uber.org/cadence/activity"
 	"go.uber.org/cadence/workflow"
-	"go.uber.org/zap"
 	"time"
 )
 
+var (
+	errArchivalUploadActivityGetDomain           = cadence.NewCustomError(errArchivalUploadActivityGetDomainStr)
+	errArchivalUploadActivityNextBlob            = cadence.NewCustomError(errArchivalUploadActivityNextBlobStr)
+	errArchivalUploadActivityConstructKey        = cadence.NewCustomError(errArchivalUploadActivityConstructKeyStr)
+	errArchivalUploadActivityBlobExists          = cadence.NewCustomError(errArchivalUploadActivityBlobExistsStr)
+	errArchivalUploadActivityMarshalBlob         = cadence.NewCustomError(errArchivalUploadActivityMarshalBlobStr)
+	errArchivalUploadActivityConvertHeaderToTags = cadence.NewCustomError(errArchivalUploadActivityConvertHeaderToTagsStr)
+	errArchivalUploadActivityWrapBlob            = cadence.NewCustomError(errArchivalUploadActivityWrapBlobStr)
+	errArchivalUploadActivityUploadBlob          = cadence.NewCustomError(errArchivalUploadActivityUploadBlobStr)
+	errDeleteHistoryActivityDeleteFromV2         = cadence.NewCustomError(errDeleteHistoryActivityDeleteFromV2Str)
+	errDeleteHistoryActivityDeleteFromV1         = cadence.NewCustomError(errDeleteHistoryActivityDeleteFromV1Str)
+)
+
 // SystemWorkflow is the system workflow code
 func SystemWorkflow(ctx workflow.Context) error {
-	id := workflow.GetInfo(ctx).WorkflowExecution.ID
-	logger := workflow.GetLogger(ctx)
-	scope := workflow.GetMetricsScope(ctx).Tagged(map[string]string{SystemWorkflowIDTag: id})
-	ch := workflow.GetSignalChannel(ctx, SignalName)
-
-	logger.Info("started new system workflow")
+	sysWorkflowInfo := workflow.GetInfo(ctx)
+	logger := globalLogger.WithFields(bark.Fields{
+		logging.TagWorkflowExecutionID: sysWorkflowInfo.WorkflowExecution.ID,
+		logging.TagWorkflowRunID:       sysWorkflowInfo.WorkflowExecution.RunID,
+	})
+	logger.Info("started system workflow")
+	globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerWorkflowStarted)
+	sw := globalMetricsClient.StartTimer(metrics.SystemWorkflowScope, metrics.SysWorkerContinueAsNewLatency)
+	ch := workflow.GetSignalChannel(ctx, signalName)
 	signalsHandled := 0
-	for ; signalsHandled < SignalsUntilContinueAsNew; signalsHandled++ {
+	for ; signalsHandled < signalsUntilContinueAsNew; signalsHandled++ {
 		var signal signal
 		if more := ch.Receive(ctx, &signal); !more {
-			scope.Counter(ChannelClosedUnexpectedlyError).Inc(1)
-			logger.Error("cadence channel was unexpectedly closed")
 			break
 		}
-		selectSystemTask(scope, signal, ctx, logger)
+		globalMetricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerReceivedSignal)
+		selectSystemTask(signal, ctx, logger, globalMetricsClient)
```

Comment:
Can we do this in its own goroutine? Otherwise each archival will be serialized on after another.

Reply:
Can we chat about this one over zoom, I want to understand this better.